test/common/http/conn_manager_impl_test.cc

```diff
@@ -4204,5 +4204,61 @@ TEST_F(HttpConnectionManagerImplTest, DisableKeepAliveWhenDraining) {
   Buffer::OwnedImpl fake_input;
   conn_manager_->onData(fake_input, false);
 }
+
+// Tests the case where after streams are closed, connection manager keeps
+// processing data, and referring to the closed streams causes crash.
+TEST_F(HttpConnectionManagerImplTest, ResetAllStreamsBeforeConnectionClose) {
+  setup(false, "envoy-custom-server", false);
+  filter_callbacks_.connection_.dispatcher_.delete_without_delay_ = true;
+  std::string input1 = "1234";
+  std::string input2 = "123456";
+
+  std::shared_ptr<MockStreamDecoderFilter> filter(new NiceMock<MockStreamDecoderFilter>());
+
+  EXPECT_CALL(*filter, setDecoderFilterCallbacks(_)).Times(1);
+
+  EXPECT_CALL(filter_factory_, createFilterChain(_))
+      .Times(1)
+      .WillRepeatedly(Invoke([&](FilterChainFactoryCallbacks& callbacks) -> void {
+        callbacks.addStreamDecoderFilter(filter);
+      }));
+
+  EXPECT_CALL(filter_callbacks_.connection_.dispatcher_, deferredDelete_(_)).Times(1);
+
+  StreamDecoder* decoder = nullptr;
+  NiceMock<MockStreamEncoder> encoder;
+
+  EXPECT_CALL(*codec_, dispatch(_))
+      .Times(2)
+      .WillRepeatedly(Invoke([&](Buffer::Instance& data) -> void {
+        if (data.length() == input1.size()) {
+          // Simulate the case where nghttp2 callback functions return error. In
+          // this case, resetAllStreams() is called, and ActiveStreams will be
+          // deleted.
+          decoder = &conn_manager_->newStream(encoder);
+          HeaderMapPtr headers{
+              new TestHeaderMapImpl{{":authority", "host"}, {":path", "/"}, {":method", "GET"}}};
+          decoder->decodeHeaders(std::move(headers), false);
+          throw CodecProtocolException("error");
+
+        } else {
+          // If connection manager keeps receiving and parsing data, referring
+          // to the deleted stream will cause crash.
+          ASSERT(data.length() == input2.size());
+          Buffer::OwnedImpl fake_data("hello");
+      // ASSERT_DEATH doesn't work in MacOS.
+#ifndef __APPLE__
+          ASSERT_DEATH(decoder->decodeData(fake_data, false), "");
```

Comment:
I thought we don't crash anymore and that was the point of this change with the tests? What's the reasoning behind having a test that doesn't pass on OSX? I'm pretty sure death tests work there, so there is there some other reason it doesn't work?

Reply:
I see your point. This test is to show after nghttp2 callback functions return error, decoder is destroyed. If the connection manager keeps processing data, envoy can crash. The fix is to set decoder to be nullptr when callbacks return error. I agree the test doesn't cover the fix, and it just shows the root cause. I am OK to revert the test.

How about we merge the metadata change first (I will revert the crash fix in that PR as well as the metadata crash test), and after that, I will submit the crash fix and the crash test as a followup PR so it's easy to track what problem it fixes? Thanks!

Reply:
Works for me!

Reply:
Sounds good.